File: go/tests/buildfarm_test.go
// buildfarm_test.go
//
// Verifies the CircleAI.BuildFarm port (buildfarm.go): agent pool
// acquire/release/list (busy-tracking), job runner state machine (Running ->
// Succeeded/Failed via Complete), artifact save/get, and null impls.

package circleai_test

import (
	"context"
	"testing"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestBuildFarm_AgentPoolAcquireRelease(t *testing.T) {
	p := circleai.NewInMemoryBuildAgentPool()
	p.Register(circleai.BuildAgent{AgentID: "l1", Kind: circleai.BuildAgentLinux})
	p.Register(circleai.BuildAgent{AgentID: "l2", Kind: circleai.BuildAgentLinux})
	p.Register(circleai.BuildAgent{AgentID: "m1", Kind: circleai.BuildAgentMac})

	a, ok := p.Acquire(context.Background(), circleai.BuildAgentLinux)
	if !ok || a.AgentID != "l1" {
		t.Fatalf("first acquire = %+v ok=%v (want l1)", a, ok)
	}
	b, ok := p.Acquire(context.Background(), circleai.BuildAgentLinux)
	if !ok || b.AgentID != "l2" {
		t.Fatalf("second acquire = %+v ok=%v (want l2)", b, ok)
	}
	if _, ok := p.Acquire(context.Background(), circleai.BuildAgentLinux); ok {
		t.Fatalf("no free Linux agents should remain")
	}
	if err := p.Release(context.Background(), "l1"); err != nil {
		t.Fatalf("release: %v", err)
	}
	if a2, ok := p.Acquire(context.Background(), circleai.BuildAgentLinux); !ok || a2.AgentID != "l1" {
		t.Fatalf("released agent should re-acquire: %+v ok=%v", a2, ok)
	}
	if lst, _ := p.List(context.Background()); len(lst) != 3 {
		t.Fatalf("list = %d, want 3", len(lst))
	}
}

func TestBuildFarm_JobRunnerStateMachine(t *testing.T) {
	r := circleai.NewInMemoryBuildJobRunner()
	job, err := r.Start(context.Background(), "l1", "repo", "main")
	if err != nil || job.Phase != circleai.BuildJobRunning {
		t.Fatalf("start = %+v err=%v", job, err)
	}
	if err := r.Complete(job.JobID, true); err != nil {
		t.Fatalf("complete: %v", err)
	}
	got, ok := r.Get(context.Background(), job.JobID)
	if !ok || got.Phase != circleai.BuildJobSucceeded {
		t.Fatalf("job after complete = %+v", got)
	}
	if err := r.Complete("ghost", true); err == nil {
		t.Fatalf("complete unknown job must error")
	}
	if _, err := r.Start(context.Background(), "", "r", "b"); err == nil {
		t.Fatalf("blank agent must error")
	}
}

func TestBuildFarm_ArtifactStore(t *testing.T) {
	s := circleai.NewInMemoryBuildArtifactStore()
	if err := s.Save(context.Background(), circleai.BuildArtifact{ArtifactID: "a1", JobID: "j", Name: "out.zip", Payload: []byte{1, 2, 3}}); err != nil {
		t.Fatalf("save: %v", err)
	}
	got, ok := s.Get(context.Background(), "a1")
	if !ok || len(got.Payload) != 3 {
		t.Fatalf("get artifact = %+v ok=%v", got, ok)
	}
	if err := s.Save(context.Background(), circleai.BuildArtifact{ArtifactID: ""}); err == nil {
		t.Fatalf("blank artifact id must error")
	}
}

func TestBuildFarm_NullImpls(t *testing.T) {
	if _, ok := circleai.NullBuildAgentPoolInstance.Acquire(context.Background(), circleai.BuildAgentLinux); ok {
		t.Fatalf("null pool must not acquire")
	}
	job, _ := circleai.NullBuildJobRunnerInstance.Start(context.Background(), "a", "r", "b")
	if job.Phase != circleai.BuildJobFailed {
		t.Fatalf("null runner job = %+v", job)
	}
}

File: typescript/tests/buildfarm.test.ts
// buildfarm.test.ts
// Verifies the CircleAI.BuildFarm port: agent-pool acquire/release, the job-runner
// state machine, the artifact store, and Null* defaults.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  BuildAgentKind,
  BuildJobPhase,
  InMemoryBuildAgentPool,
  InMemoryBuildJobRunner,
  InMemoryBuildArtifactStore,
  NullBuildAgentPool,
  NullBuildJobRunner,
  buildAgent,
  buildArtifact,
} from "../src/buildfarm/index";

describe("InMemoryBuildAgentPool", () => {
  it("acquires a free agent of a kind and releases it", async () => {
    const pool = new InMemoryBuildAgentPool();
    pool.register(buildAgent("l1", BuildAgentKind.Linux, "ubuntu", null));
    pool.register(buildAgent("m1", BuildAgentKind.Mac, "sonoma", "M2"));
    const a = await pool.acquireAsync(BuildAgentKind.Linux);
    assert.equal(a?.agentId, "l1");
    // No more Linux agents free.
    assert.equal(await pool.acquireAsync(BuildAgentKind.Linux), null);
    await pool.releaseAsync("l1");
    assert.equal((await pool.acquireAsync(BuildAgentKind.Linux))?.agentId, "l1");
    assert.equal((await pool.listAsync()).length, 2);
  });
});

describe("InMemoryBuildJobRunner", () => {
  it("starts a Running job and completes it", async () => {
    const runner = new InMemoryBuildJobRunner();
    const job = await runner.startAsync("l1", "repo", "main");
    assert.equal(job.phase, BuildJobPhase.Running);
    runner.complete(job.jobId, true);
    assert.equal((await runner.getAsync(job.jobId))?.phase, BuildJobPhase.Succeeded);
  });

  it("marks a job Failed", async () => {
    const runner = new InMemoryBuildJobRunner();
    const job = await runner.startAsync("l1", "repo", "dev");
    runner.complete(job.jobId, false);
    assert.equal((await runner.getAsync(job.jobId))?.phase, BuildJobPhase.Failed);
  });

  it("throws completing an unknown job", () => {
    assert.throws(() => new InMemoryBuildJobRunner().complete("nope", true));
  });
});

describe("InMemoryBuildArtifactStore", () => {
  it("saves and fetches an artifact", async () => {
    const store = new InMemoryBuildArtifactStore();
    await store.saveAsync(buildArtifact("art1", "job1", "app.apk", new Uint8Array([1, 2])));
    const got = await store.getAsync("art1");
    assert.equal(got?.name, "app.apk");
    assert.equal(await store.getAsync("missing"), null);
  });
});

describe("Null build-farm defaults", () => {
  it("NullBuildAgentPool acquires nothing", async () => {
    assert.equal(await NullBuildAgentPool.instance.acquireAsync(BuildAgentKind.Ios), null);
  });
  it("NullBuildJobRunner returns a failed job", async () => {
    const job = await NullBuildJobRunner.instance.startAsync("a", "r", "b");
    assert.equal(job.phase, BuildJobPhase.Failed);
  });
});
